packages/next/src/server/api-utils/index.ts

```diff
@@ -20,19 +20,33 @@ export type __ApiPreviewProps = {
   previewModeSigningKey: string
 }
 
-export function wrapApiHandler<T extends (...args: any[]) => any>(
-  page: string,
-  handler: T
-): T {
+export function wrapApiHandler<
+  T extends (
+    ...args: [req: IncomingMessage, res: ServerResponse, ...any[]]
+  ) => any,
+>(page: string, handler: T): T {
   return ((...args) => {
     getTracer().setRootSpanAttribute('next.route', page)
     // Call API route method
     return getTracer().trace(
       NodeSpan.runHandler,
       {
         spanName: `executing api route (pages) ${page}`,
+        manualSpanEnd: true,
       },
-      () => handler(...args)
+      (span) => {
+        if (span) {
+          const res = args[1]
+          res.end = new Proxy(res.end, {
```

Comment:
Is there no `res.on('end', ...)`?

Reply:
Correct me if I am wrong, but res.on('end') will not work on Vercel because the underlying lambdas freeze on res.end() which happens before res.on('end').

If that is not the case anymore it would be amazing because it would simplify the code here a lot.